### PR TITLE
Update `UserMailer` to use paramaterization

### PR DESCRIPTION
**Pull request by mjankowski**
Similar to previous PRs for `NotificationMailer` and `AdminMailer`.

Summary:

- Adds missing spec coverage for the welcome and backup_ready mailers
- Add before_action for setting common instance vars
- Add default `to` value
- Add parameterization to almost all mailers (did not do the devise-descended ones in this PR)
- Use `default_i18n_subject` where it works, or a custom devise version where it doesn't (the built-in one does not pick up the right i18n scope from parent classes. Might revisit this in future PR).
- Fix one bug (I think!?) where the appeal_rejected mailer was not actually being sent

### CODE REVIEW

File: app/controllers/admin/disputes/appeals_controller.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def reject
     authorize @appeal, :approve?
     log_action :reject, @appeal
     @appeal.reject!(current_account)
-    UserMailer.appeal_rejected(@appeal.account.user, @appeal)
+    UserMailer.with(user: @appeal.account.user).appeal_rejected(@appeal).deliver_later
```
Comment (mjankowski) [inline app/controllers/admin/disputes/appeals_controller.rb:23]: The spec for this action was only verifying that `:appeal_rejected` was called, not that delivery_later was called. This seems like a bugfix?

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -1,20 +1,35 @@
 # frozen_string_literal: true
 
 class UserMailer < Devise::Mailer
+  DEVISE_ACTIONS = [
```
Comment (mjankowski) [inline app/mailers/user_mailer.rb:4]: This was the least-bad thing I could think of to highlight the fact that the devise-overriding actions aren't getting the same changes as the others.

I think there's room for future refactor here to keep all the devise ones in their own mailer and move the rest to other mailer(s), but didn't want this PR to get even bigger.

**Comment by ClearlyClaire** [comment]
Doesn't this change the signature of the jobs? In which case already-enqueued jobs won't be able to be processed by the new code. Is there a way we can handle that?

**Comment by mjankowski** [comment]
> Doesn't this change the signature of the jobs?

Probably.

> In which case already-enqueued jobs won't be able to be processed by the new code. Is there a way we can handle that?

It probably depends on the instance, but if we assume a busy instance with a massive queue of jobs that will be slow to actually clear, and we want zero downtime deploy, the most conservative thing would be:

- Update this PR to introduce new class(es) instead of changing the existing signatures. Update all code to call new class instead.
- Deploy that, leaving old definitions in place so old queues can run down and new queues start to be used
- Once cleared, eventually remove old definitions
- Optionally, rename new classes back to original class names - making sure to repeat the process of leaving the new-named classes in place for a deploy cycle so that any of those that are still in queues can send correctly before removing

**Comment by ClearlyClaire** [comment]
Hm, yeah, I'm not sure this is worth the change. The other changes seem fine though!

**Comment by mjankowski** [comment]
> Hm, yeah, I'm not sure this is worth the change. The other changes seem fine though!

Fair enough. The only other ideas I have re: parameterization, are:

- I had planned on doing a followup refactor after this one was merged to separate some of these specific emails into separate mailers. At minimum, all the devise-related ones are different enough than all the custom ones that doing that separation seemed logical. Beyond that I was going to review how it looked. If we went this route, that would also be a justification for changing the parameterization during that change.
- It might be possible to leave the previous method signatures in places without changing the class, and just have them delegate to the new style. I'll look into this and report back.

Regardless of either of those, I'll pull the unrelated changes out into other PRs.

**Comment by mjankowski** [comment]
Rebased this after the merging of the other PR which added the non-parameterization stuff.

I see that I missed 1-2 changes which were around spec coverage and not narrowly related to parameterization. Will pull those out and do another PR.

For the parameterization stuff, I'm concluding that any changes which change the method signature for queueing purposes are deemed just not worth it migration-wise? If that's the case, I think there are 1-2 more things I can pull out of this which may be useful, and then I can close this and drop the signature-changing portoins.

**Comment by mjankowski** [comment]
Closing this one due to the migration issues discussed. If we do want to go this route there's a more viable migration path discussed in other PRs.

**Pull request closed without merging**